Removes laminas-json as dependency and uses PHP’s native json functions

mairo744 opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

This change removes all usage of `Laminas\Json` in favor of PHP’s native `json_encode()` and `json_decode()` functions.

**Reason for change:**
- `laminas/laminas-json` is abandoned and no longer maintained.

**Changes made:**
- Replaced `LaminasJson::encode()` calls with `json_encode()` using `JSON_THROW_ON_ERROR`.
- Replaced `LaminasJson::decode()` calls with `json_decode()` and `$associative` mapping from the configured object decode type.
- Preserved existing cycle check logic and exception wrapping for backward-compatible error handling.

**Impact:**
- Removes dependency on abandoned `laminas/laminas-json` package.
- Potentially changes exception types internally (now using `JsonException` from PHP core), but they are wrapped to maintain public API compatibility.
- This is a BC break

mmalac committed:
feat remove abandoned laminas json with native php functions

Signed-off-by: mmalac <mmalac@sygic.com>

fezfez commented:
@froschdesign any news ?

samsonasik commented on src/Adapter/JsonOptions.php:
```diff
@@ -37,30 +34,15 @@ public function getEnableJsonExprFinder(): bool
      * @throws Exception\InvalidArgumentException
      */
-    public function setObjectDecodeType(int $type): void
+    public function setAssocArray(bool $assocArray): void
```
public method change seems bc break

fezfez replied:
@samsonasik could you open 4.0.x, @mairo744 pointed that this is BC.

samsonasik replied:
I've created 4.0.x branch https://github.com/laminas/laminas-serializer/tree/4.0.x

fezfez replied:
@mairo744 could you target 4.0.x ?

gsteel commented on src/Adapter/JsonOptions.php:
```diff
@@ -37,30 +34,15 @@ public function getEnableJsonExprFinder(): bool
      * @throws Exception\InvalidArgumentException
      */
-    public function setObjectDecodeType(int $type): void
+    public function setAssocArray(bool $assocArray): void
```
Note that in v3, we should add:

```php
/**
 * @deprecated Use the `setAssocArray` method
 * @see setAssocArray()
 */
public function setObjectDecodeType(int $type): void
{
    // ...
}

public function setAssocArray(bool $flag): void
{
    $this->setObjectDecodeType($flag ? LaminasJson::TYPE_ARRAY : LaminasJson::TYPE_OBJECT);
}
```

mairo744 replied:
https://github.com/laminas/laminas-serializer/pull/89

gsteel commented on src/Adapter/JsonOptions.php:
```diff
@@ -37,30 +34,15 @@ public function getEnableJsonExprFinder(): bool
     }
```
These methods (And related property) should be removed because they are specific to `Laminas\Json`.

They should also be deprecated in a patch to v3

gsteel commented on src/Adapter/Json.php:
```diff
@@ -72,13 +81,54 @@ public function serialize(mixed $value): string
+    /**
+     * @param mixed[] $options
+     */
+    private function encode(mixed $value, bool $cycleCheck, array $options): string
```
I doubt there are existing tests that cover this because it was previously handled by `Laminas\Json`. Can you verify these lines are covered or write new tests please.

mairo744 replied:
I have added some new tests

boesing commented on src/Adapter/Json.php:
```diff
@@ -72,13 +81,54 @@ public function serialize(mixed $value): string
+            : 0;
+
+        $encoded = json_encode($value, $jsonOptions | JSON_THROW_ON_ERROR);
+        if ($encoded === false) {
```
unreachable code. PHP will already throw due to `JSON_THROW_ON_ERROR` so `$encoded` will never contain `false`.

I wonder why psalm is not yelling here - maybe because it does not parse the JSON options accordingly 🤔

mmalac committed:
refactor: remove unused JsonOptions and simplify Json adapter implementation

- Replaced custom logic with native JSON functions, utilizing `JSON_UNESCAPED_UNICODE` and `JSON_UNESCAPED_SLASHES`.
- Removed unsupported options such as `cycle_check` and `enable_json_expr_finder`.
- Add tests for serialization and deserialization

Signed-off-by: mmalac <mmalac@sygic.com>

mmalac committed:
docs: add adapter documentation for v4

Signed-off-by: mmalac <mmalac@sygic.com>

mairo744 commented:
I have made the requested changes. 
Thanks again for feedbacks.
Could you please take another look when you have a moment?

froschdesign commented:
@samsonasik 
Do you have any further comments on this?

samsonasik reviewed (approved):
LGTM for 4.0.x branch 👍

froschdesign commented:
@mairo744 
Thank you for your time and this contribution! 👍
